## ---------------------------
##
## Script name: 3_PhenomeWide_GWAS.R
##
## Purpose of script: General GWAS execution and analysis for all traits.
##
## Author: M.Sc. Robin Lindner
##
## Date Created: 2025-07-08
##
## Copyright (c) Robin Lindner, 2025
## Email: robin.lindner@uni-potsdam.de
##
## ---------------------------
##
## Notes:
##   1. Normalized BLUPs were used as the phenotype.
##   2. FarmCPU was used as the method with the first 4 PC of the GRM as additional
##      covariates.
##   3. The SimpleM algorithm was used to determine the number of effective 
##      markers, which were subsequently used to adjust for multiple testing 
##      with the Bonferronni correction.
##   4. SNP Positions in the GWAS results might not match with the ones mentioned
##      in the main text, since the genotyping data was generated using an outdated
##.     reference assembly. The SNPs were mapped to a current assembly. 
##.     The mapping is stored in SNP_remap.csv .
##
## ---------------------------

## load up our functions into memory:

source("0_utils.R")
## ---------------------------  

## ---- HPC_GWAS.R: ----
## GWAS computation can be outsourced and executed with HPC_GWAS.R for the 
## normalized BLUP file generated in 2_GenomicEffectBLUPs_Heritability.R

dir.create("../Data/Generated/GWAS_results")
dir.create("../Data/Generated/GWAS_support")
## run: 
## Rscript HPC_GWAS.R ../Data/Generated/BLUPs_normalized.csv ../Data/Generated/GWAS_results ../Data/Genotype/B1K_final.vcf ../Data/Genotype/B1K_final_GRM.csv ../Data/Generated/GWAS_support


## ---- Multiple testing correction ----
source("simpleM.R")
Meff=c()
Meff[1] = computeMeff("../Data/Genotype/B1K_final_numeric_chr1.txt")
Meff[2] = computeMeff("../Data/Genotype/B1K_final_numeric_chr2.txt")
Meff[3] = computeMeff("../Data/Genotype/B1K_final_numeric_chr3.txt")
Meff[4] = computeMeff("../Data/Genotype/B1K_final_numeric_chr4.txt")
Meff[5] = computeMeff("../Data/Genotype/B1K_final_numeric_chr5.txt")
Meff[6] = computeMeff("../Data/Genotype/B1K_final_numeric_chr6.txt")
Meff[7] = computeMeff("../Data/Genotype/B1K_final_numeric_chr7.txt")

MeffTotal = sum(Meff)
sig_threshold = 0.05/MeffTotal

## ---- HPC_ESA.R: ----
## This script extracts significant associations for all 1456 GWAS generated by
## HPC_GWAS.R

## run:
## Rscript HPC_ESA.R ../Data/Generated/GWAS_results [sig_threshold] ../Data/Generated/significant_associations.csv


## ---- General results ----

snp_map = read.csv(geno_remap_file,row.names = 1)
sig_associations = read.csv(sig_associations_file)
sig_associations$Group = trait_groups$Group[match(sig_associations$Trait,trait_groups$Trait)]


## Trait associations per marker
traits_per_marker = sig_associations %>%
  group_by(SNP) %>%
  summarize(TraitNr = length(unique(Trait)))

## Data frame storing group specific measurement numbers
group_measurement_stats = data.frame(Group=unique(trait_groups$Group),
                                     nTrait = c(9,41,11,2,2,41,6,6,3),
                                     nDAT = c(19,9,35,40,19,9,6,6,1))
group_measurement_stats = group_measurement_stats %>%
  mutate(measurements = nTrait*nDAT)

## Data frame storing general association statistics
# | Group | number of time points where associations were found | total number of found associations | number of unique SNP in those associations | number of Measurements | number of association per measurement |
all_assoc_group_stats = sig_associations %>%
  group_by(Group) %>%
  summarize(nDAT = length(unique(DAT)),
            nAssoc = n(),
            nSNP = length(unique(SNP))) %>%
  mutate(nMeasure = group_measurement_stats$measurements[match(Group,group_measurement_stats$Group)],
         assoc_per_measure = nAssoc/group_measurement_stats$measurements[match(Group,group_measurement_stats$Group)])


## ---- Jaccard similarity of SNP sets between traits ----
nTrait = 121
jac_matrix = matrix(0,nrow = nTrait,ncol=nTrait)
for(i in 1:(nTrait-1)){
  snp_trait_i = sig_associations%>%filter(Trait==trait_groups$Trait[i])%>%dplyr::select(SNP)%>%distinct(.keep_all = T)
  if(nrow(snp_trait_i)==0){next}
  for(j in (i+1):nTrait){
    snp_trait_j = sig_associations%>%filter(Trait==trait_groups$Trait[j])%>%dplyr::select(SNP)%>%distinct(.keep_all = T)
    if(nrow(snp_trait_j)==0){next}
    jac_matrix[i,j] = jaccard(snp_trait_i$SNP,snp_trait_j$SNP)
  }
}
jac_matrix[lower.tri(jac_matrix,diag = T)]=NA
rownames(jac_matrix) = colnames(jac_matrix) = trait_groups$Trait

jac_long <- as.data.frame(jac_matrix) %>%
  tibble::rownames_to_column("Trait1") %>%
  pivot_longer(cols = c(-1),names_to = "Trait2",values_to = "Value") %>%
  filter(!is.na(Value)) %>%
  mutate(WithinGroup = sameGroup(Trait1,Trait2))
  
# Wilcoxon rank sum test for difference in mean Jaccard similarity between and within groups
jaccard_between_groups = jac_long$Value[!jac_long$WithinGroup]
mean(jaccard_between_groups)
sd(jaccard_between_groups)
jaccard_within_groups = jac_long$Value[jac_long$WithinGroup]
mean(jaccard_within_groups)
sd(jaccard_between_groups)

wilcox.test(jaccard_between_groups,jaccard_within_groups)

## ---- Jaccard similarity of SNP sets between groups ----

nGroup = 9
groups = unique(trait_groups$Group)
jac_matrix = matrix(0,nrow = nGroup,ncol=nGroup)

for(i in 1:(nGroup-1)){
  snp_trait_i = sig_associations%>%filter(Group==groups[i])%>%dplyr::select(SNP)%>%distinct(.keep_all = T)
  if(nrow(snp_trait_i)==0){next}
  for(j in (i+1):nGroup){
    snp_trait_j = sig_associations%>%filter(Group==groups[j])%>%dplyr::select(SNP)%>%distinct(.keep_all = T)
    if(nrow(snp_trait_j)==0){next}
    jac_matrix[i,j] = jaccard(snp_trait_i$SNP,snp_trait_j$SNP)
  }
}
jac_matrix[lower.tri(jac_matrix,diag = T)]=NA
rownames(jac_matrix) = colnames(jac_matrix) = groups

jac_long_group <- as.data.frame(jac_matrix) %>%
  tibble::rownames_to_column("Group1") %>%
  pivot_longer(cols = c(-1),names_to = "Group2",values_to = "Value") %>%
  filter(!is.na(Value))

jac_long_group
mean(jac_long_group$Value)
sd(jac_long_group$Value)


## ---- Trait-group specific distribution of SNP across genome ----

## Compute the SNP density [SNP/Mb]
remap = read.csv(geno_remap_file,row.names = 1)
SNP_density = remap %>%
  group_by(Chromosome) %>%
  summarize(Count = n(),
            Length = max(new_position)) %>%
  mutate(Density = (Count/Length)*1000000)



## Group enrichment statistics:
## | Group | Chromosome | sig. SNP / number of total SNP on chromosome | Enrichment | standardized residuals of chi^2 test
group_enrichment_tab = data.frame(Group = NA,Chrom=NA,normSNP=NA,Enrichment=NA,stdRes=NA)
## Enrichment is calculated as:
## proportion of SNP on chromosome that are significant / proportion of SNP on genome that are significant


## Results of Chi^2 test for homogeneous distribution of significant SNP across all chromosomes.
group_chisq_tab = data.frame(Group=NA,chisq.p=NA,nSNP=NA)

k=1
l=1
for(group in groups){
  count_mat = data.frame(Chrom=NA,Associated=NA,NotAssociated=NA,Total=NA)
  for(chrom in 1:7){
    temp = sig_associations %>%
      filter(Group==group & CHROM==chrom)
    nSNP = length(unique(temp$SNP))
    normalized_SNP = nSNP / SNP_density$Count[chrom]
    enrichment = normalized_SNP / (length(unique(all_assoc$SNP)) / sum(SNP_density$Count))
    count_mat[chrom,] = c(chrom,nSNP,SNP_density$Count[chrom]-nSNP,SNP_density$Count[chrom])
    group_enrichment_tab[k,] = c(group,chrom,normalized_SNP,enrichment,NA)
    k=k+1
  }
  t=chisq.test(count_mat[,-c(1,4)])
  group_chisq_tab[l,] = c(group,t$p.value,sum(count_mat[,2]))
  l=l+1
  group_enrichment_tab[(k-7):(k-1),5] = t$stdres[,1]
  
}
write.csv(group_enrichment_tab[,c(1,2,5)],"../Supplements/assoc_chrom_enrichment.csv",row.names = F)


## ---- Trait-group specific temporal stability of SNP associations ----
group_measurement_stats

unique_SNP = sig_associations %>% 
  group_by(Trait) %>%
  summarize(uSNP = length(unique(SNP)))


association_nrs=sig_associations %>%
                  group_by(Trait,Group,DAT) %>%
                  summarize(Count=n()) %>%
                  group_by(Trait,Group) %>%
                  summarize(mean = mean(Count),
                            sd =sd(Count)) 

association_nrs$expectedAssoc = unique_SNP$uSNP / group_measurement_stats$nDAT[match(association_nrs$Group,group_measurement_stats$Group)]
association_nrs$TemporalStability = association_nrs$mean / association_nrs$expectedAssoc

temporal_stability_groups = association_nrs %>%
  group_by(Group) %>%
  summarise(Temp_stability_Mean = mean(TemporalStability),
            Temp_stability_sd = sd(TemporalStability))

## ---- effects of pSNP ----##

pSNP=traits_per_marker %>%
  arrange(desc(TraitNr)) %>%
  head(3)

# Takes SNP ID and data.frame containing significant associations as inputs.
# Uses these in conjunction to select the traits for which the comparison in normalized
# BLUP values should be made. 
pSNP1 = createTraitcomparisionDF(pSNP$SNP[1],sig_associations)

# Data.frame of the normalized BLUP values used in the comparison
pSNP1[[1]]

# Data.frame of TukeyHSD test results between normalized BLUP values for genotypes
# with differentiating allele states at the SNP position.
t1=pSNP1[[2]]




pSNP2 = createTraitcomparisionDF(pSNP$SNP[2],sig_associations)

pSNP3 = createTraitcomparisionDF(pSNP$SNP[3],sig_associations)

# Compute overdominance
od_df = pSNP3[[2]] %>% group_by(Trait) %>% summarize(Overdominance = (Diff[Comparison=="Homozygotic major-Heterozygotic"]<0 && Diff[Comparison=="Homozygotic minor-Heterozygotic"]<0))



## ---- Candidate gene inference ---- 
# see file:
 
